### PR TITLE
Feature/traces

### PR DESCRIPTION
**Pull request by Jacarte**
### Log format

```

<trace> -> <log> \n <trace> | <EOF>

<log> -> <event_type>,<time>,<args>

<time> -> \d+
<event_type> -> "msize" | "mread" | "mwrite" | "mgrow" | "spush" | "spop"

<args> -> <arg>,<args> | <arg>

<arg> -> \w+

```

### Example

``` mread,1567590525454,i32,100,10, ```

This log means, ```memory read at time 1567590525454, read as 4 bytes integer(i32), in address 100, result  value 10 ```

### Filtering logs

You can set a filter for event logs based on the first element in the line. The filter can be specified as a regular expression in the ```reference.conf``` file. For example, to log only memory operations: ```filter = "mread|mwrite"```.

### Tracers
- StdTracer: Prints all event lines in the console
- InFileTracer: Writes all events to log files. The name of log file and the destination folder are in the ```reference.conf``` file, inside the tracer property declaration. 

To change the tracer, you need to specify the **Java class name** in the tracer properties inside the file ```reference.conf```, i.e. ```tracerName = "swam.runtime.internals.tracer.StdTracer"```

To specify no-tracer, you must set the ```tracerName``` property to ```"None"```

Default tracer: ```"None"```

### Output example

```
mread,446479078,i32,0,0
mread,451988189,i32,0,2
mwrite,452499619,i32,0,2
mread,452918915,i32,0,2
mwrite,454010692,i32,65536,3
mwrite,447102186,i32,0,2
mread,449056814,i32,0,2
mread,453380464,i32,65536,0
mread,454471107,i32,65536,3

```

### CODE REVIEW

**Comment by satabin** [comment]
I will have a look at it and give you hints on where code can be improved. Can you also please enable travis.org on your repository, so that we can have a build of this PR? Thanks.

File: runtime/src/swam/runtime/Interface.scala [hunk 1]
```diff
@@ -206,3 +206,53 @@ trait Memory[F[_]] extends Interface[F, MemType] {
...
+    *  $boundaries
+    */
+    def traceEvent(eventName: String, args: Any*): Unit
+  
+  
+    /** Make a background request on the event record call
+   
+    *  $boundaries
+    */
+    protected def executeOnBack(f: () => Unit) = {
```
Comment (satabin) [inline runtime/src/swam/runtime/Interface.scala]: Hmm that one is tricky to explain in a review, I will make a suggestion using the cats-effect stuffs instead.

File: runtime/src/swam/runtime/internals/instance/Instantiator.scala [hunk 2]
```diff
@@ -31,7 +31,7 @@ import runtime._
 
 import scala.language.higherKinds
 
-private[runtime] class Instantiator[F[_]](engine: Engine[F])(implicit F: Async[F]) {
+private[runtime] class Instantiator[F[_]](engine: Engine[F])(implicit F: Async[F], tracer: Tracer) {
```
Comment (satabin) [inline runtime/src/swam/runtime/internals/instance/Instantiator.scala]: Why make the `Tracer` an implicit argument?

File: runtime/src/swam/runtime/internals/instance/MemoryInstance.scala [hunk 3]
```diff
@@ -40,47 +40,91 @@ private[runtime] class MemoryInstance[F[_]](min: Int, max: Option[Int], onHeap:
...
-  def writeByte(idx: Int, v: Byte) =
+  def writeByte(idx: Int, v: Byte) = {
+    if (tracer != null)tracer.traceEvent("msize", size)
     F.delay(buffer.put(idx, v))
+  }
 
-  def readByte(idx: Int) =
-    F.delay(buffer.get(idx))
+  def readByte(idx: Int) = {
+    F.delay({
```
Comment (satabin) [inline runtime/src/swam/runtime/internals/instance/MemoryInstance.scala]: style: you can drop the parentheses around the braces, which makes it look nicer:
```scala
F.delay {
  ...
}
```

File: runtime/src/swam/runtime/internals/tracer/InFileTracer.scala [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import config._
+import java.io._
+import java.text.SimpleDateFormat
+import java.util._
+import java.util.UUID.randomUUID
+import scala.collection.mutable.ListBuffer
+/** Write events log file
+ * Every log file is grouped by the engine execution scope
+  */
+class InFileTracer(val conf: EngineConfiguration) extends Tracer{
```
Comment (satabin) [inline runtime/src/swam/runtime/internals/tracer/InFileTracer.scala]: For this I think it would be better to use the `fs2-io` library to write the files.

**Comment by Jacarte** [comment]
On it, thanks for the reviews

**Comment by Jacarte** [comment]
The tracer is deactivated in tests due to timeout, thats why the code is not covered. Any idea about how can we include tracers tests?

**Comment by satabin** [comment]
I have to have a deeper look at it. The IOs performed by the tracer are currently not safe, and I'd like to improve that. IO safety in Swam is based on `cats-effect` and `fs2-io` if you want to do it yourself, you might want to have a look at the [documentation of the former one](https://typelevel.org/cats-effect/datatypes/).

**Comment by Jacarte** [comment]
Sure ! I will take a look

**Comment by Jacarte** [comment]
Anyway we can create an extra test file instantiating the concrete implementations. This will remove the bad coverage report

**Comment by monperrus** [comment]
What about using a known trace format? By doing this, we would be able to plug the traces in different tools later on.

For instance, the Common Trace Format: https://diamon.org/ctf/

WDYT?

**Comment by satabin** [comment]
> What about using a known trace format? By doing this, we would be able to plug the traces in different tools later on.
> 
> For instance, the Common Trace Format: https://diamon.org/ctf/
> 
> WDYT?

I love the idea of using a standard format, so I am entirely in favor of this proposition.

**Comment by Jacarte** [comment]
Ok, on it

**Comment by satabin** [comment]
I saw you did a lot of change, I will review this this weekend. Do you prefer me to just comment and point you in the direction I mean when I see something or directly suggest the diff?

**Comment by Jacarte** [comment]
Hi Lucas, any comment will be appreciated, thanks again !. I prefer you point the direction of the solution, but, if you think that implement the solution will take too long for me, just directly suggest it.

**Comment by satabin** [comment]
So, this is a big PR, but will bring interesting learnings for both of us I guess. As a starter, you can start looking at Codacy warnings for best practices.

Idiomatic scala doesn't use `null` but instead the `Option[T]` type. I use `null` in several places in Swam for performance reasons, but I always start with Option and see if it is a bottleneck. So I would recommend for the `tracer` to be an `Option[Tracer]` first. This is a nice type as it forces you to _check for null_ at the type level. for instance, where you do

```scala
if(tracer != null) tracer.traceSomething(...)
```

With option it translates to

```scala
tracer.foreach(tracer => tracer.traceSomething(...))
// or with anonymous paramter name
tracer.foreach(_.traceSomething(...))
// or with `for` comprehension
for(tracer <- tracer)
  tracer.traceSomething(...)
```

Also, you should never use `return`. In Scala the last expression of the function is returned as return value. As a functional language, there is not statements, but everything yield a value, potentially of type `Unit` (the Scala `void`). I also tend to dislike _useless_ braces, I mean for functions or blocks with a single expression, braces are optional and I do not write them, I find it more readable.
For instance instead of
```scala
def add(x: Int, y: Int): Int = {
 x + y
}
```
I prefer this version:
```scala
def add(x: Int, y: Int): Int =
 x + y
```
Exception is for nested if blocks, or if blocks with match block in at least one branch, I find it more readable if branches have braces.

For logging, I would prefer if you use slf4j instead of explicitly java logging API, because the former one allows people to choose their backend (logback, log4j, java logging, noop, ...). You do not have to bind the `runtime` module to a specific implementation of slf4j, as it is a library, and can choose the one you like in the test module.

This is it for now for the _raw_ Scala part.

Now comes the fun part: functional Scala with cats and cats-effect!
This is the programming style I use in Swam and that I like (like… a lot, I should say), which is based on type classes, inspired by Haskell, and basically I use it a lot to make all side-effects safe. This includes all I/O (logging, reading/writing to a file, ...) and is higher level than doing it directly. This is important because all computations in Swam are suspended in the `F` type, which defers the execution as late as possible. So for instance when you have a expression of type `F[T]` it represents a computation that will yield a value of type `T` when executed. `F` usually has (at least)  a monad instance, which means nothing else than you can bind the `F[T]` to a function that will compute a new value with the result if it and yield a new `F[U]`. In Scala the best syntax for it is the `for`-comprehension, which is like the `do`-notation in Haskell and allows you to write in a sequential way the monadic bindings. For instance, assuming all functions return an `F` we can write things like:
```scala
for {
  res1 <- f(args)
  res <- g(res1)
} yield res
```
which creates a new `F` which, when executed, will execute `f`, when the result is available, use it fo execute `g` and yield the result of this. This is merely syntactic sugar that uses `flatMap` (the Scala monadic bind operator). For instance the code above is equivalent to:
```scala
f(args).flatMap { res1 =>
  g(res1)
}
```
I wrote _equivalent_ because it is not 1-to-1 translated as such by the vanilla Scala compiler but, in our case it is thanks to a compiler plugin. Anyway, even with vanilla compiler, the behavior is the same as what I described.
In your code, side-effects are executed immediately because they are not wrapped in that deferring `F`, which might in some cases lead to unexpected results. On top of that, I saw you use low-level synchronization mechanisms (like re-entrant locks) to order events. cats-effect and fs2 (streaming library based on it) offer higher-level and safer ways to do it. It makes it easier for you also to reason about your program. For instance for your trace events, the [`Queue`](https://fs2.io/concurrency-primitives.html) type from fs2 might be a good match where you can push new events and a concurrent stream will then process them without blocking execution. This part is really interesting and requires some time to get used to in general, but I trust you can handle it.

I tried to give a really short overview of the concepts involved in Swam, but for more details and explanation the cats-effect and fs2 websites have a really good introductory documentation. If you want to learn more about what is a monad, an applicative, and all these weird names, the [cats website](https://typelevel.org/cats/typeclasses) also has a description of what every of these mean.

I will anyway continue review your changes and give you more pointers if I see something that I forgot, and do not hesitate to ask in case you have any doubt.

File: runtime/src/swam/runtime/config/EngineConfiguration.scala [hunk 5]
```diff
@@ -59,14 +59,14 @@ case class LowLevelStackConfiguration(size: Information)
 
 case class HighLevelStackConfiguration(size: Information, callDepth: Int)
 
-case class TraceConfiguration(handler: HandlerType,
+case class TraceConfiguration(var handler: HandlerType,
```
Comment (satabin) [inline runtime/src/swam/runtime/config/EngineConfiguration.scala]: Why is a `var` needed and not a `val`? Prefer immutable values everywhere. #FunctionalProgrammingRocks

Comment (Jacarte) [reply runtime/src/swam/runtime/config/EngineConfiguration.scala]: Well...there is no need. I was using it to change it in the Test suite for the tracer. Thanks again !! :)

**Comment by Jacarte** [comment]
I was looking for CTF library in Java. @satabin do you know some?  Do we need to create our own supporting library for this?

**Comment by satabin** [comment]
What do you mean by CTF?

**Comment by Jacarte** [comment]
Common Trace Format https://diamon.org/ctf/

**Comment by Jacarte** [comment]
Ok, thanks, I will let you know with new changes then

**Comment by satabin** [comment]
Hi @Jacarte,
Sorry I haven't invested much time in that project lately. I will review your last changes, and I actually noticed something: I told you something wrong on where memory reads happen. This is also a reason for my ongoing rewriting of the interpreter internal structure, because even I am confused with it. The `Frame.read*` methods are used to read the instruction code, not the linear memory. This occurs in the `MemoryInstance` class. However, it is not the only class that represents a linear memory, simply the most common one. Until I refactor my interpreter, you should trace these calls. I would recommend you to write a wrapper [`Memory`](https://github.com/satabin/swam/blob/master/runtime/src/swam/runtime/Interface.scala#L114) instance, that you would use to wrap other instances in case you enable tracing. I will make it more convenient in the future.

**Comment by satabin** [comment]
So I merged my big rewrite to make things a bit easier. It removes the high/low level interpreter implementations and only leaves the low level one. The high-level interpreter was the very first attempt, and unusable in practice. The rebasing can be painful, but I can help if you give me write access to your branch, this is an option of the PR.

Also about the Codacy errors, don't worry about all of them, some are no problem in our case, I will ignore the ones that are not important.

**Comment by Jacarte** [comment]
Done, write access enabled. Thanks!

**Comment by satabin** [comment]
So, I tried to rebase and this is hell :)
I have a proposal: I take the current state of your PR, and recreate a single commit that implement it (with you as an author) and then push to your branch? Also when rebasing I found things that I didn't manage to explain properly, what do you think if I fix them in the process and we can discuss it afterwards?

**Comment by Jacarte** [comment]
> So, I tried to rebase and this is hell :)
> I have a proposal: I take the current state of your PR, and recreate a single commit that implement it (with you as an author) and then push to your branch? Also when rebasing I found things that I didn't manage to explain properly, what do you think if I fix them in the process and we can discuss it afterwards?

Yes, sure! thanks again

**Comment by satabin** [comment]
Ok, I'll try to do it tomorrow.

**Comment by Jacarte** [comment]
> Ok, I'll try to do it tomorrow.

We can work at the same time if you want. I will be available for this tomorrow also

**Comment by satabin** [comment]
Sure, we can do a remote pair programming session if you want.

**Comment by satabin** [comment]
Hi @Jacarte,

After having left this PR open for too long, and failing to find a time to work on this together, I copied over your ideas in a new branch and PR and implemented it on the new engine in #54.

The main differences are:
 - a more Scala-ish code
 - a less invasive tracer integration
 - less overhead when no tracer is given

I left you as author of the commit because it is purely based on your work. If you are fine with this, I will merge this other PR.

**Pull request closed without merging**